samcli/lib/list/testable_resources/testable_resources_producer.py

```diff
@@ -0,0 +1,367 @@
+
+
+class TestableResourcesProducer(ResourceMappingProducer, Producer):
+    def __init__(
+        self,
+        stack_name,
+        region,
+        profile,
+        template_file,
+        cloudformation_client,
+        iam_client,
+        cloudcontrol_client,
+        apigateway_client,
+        apigatewayv2_client,
+        mapper,
+        consumer,
+    ):
+        super().__init__(
+            stack_name, region, profile, template_file, cloudformation_client, iam_client, mapper, consumer
+        )
+        self.stack_name = stack_name
+        self.region = region
+        self.profile = profile
+        self.template_file = template_file
+        self.cloudformation_client = cloudformation_client
+        self.iam_client = iam_client
+        self.cloudcontrol_client = cloudcontrol_client
+        self.apigateway_client = apigateway_client
+        self.apigatewayv2_client = apigatewayv2_client
+        self.mapper = mapper
+        self.consumer = consumer
+
+    def get_function_url(self, identifier: str) -> Any:
+        """
+        Gets the function url of a Lambda Function
+
+        Parameters
+        ----------
+        identifier: str
+            The identifier or physical ID
+
+        Returns
+        -------
+        furl: str
+            The function url in the form of a string
+        """
+        try:
+            response = self.cloudcontrol_client.get_resource(TypeName="AWS::Lambda::Url", Identifier=identifier)
+            if not response.get("ResourceDescription", {}).get("Properties", {}):
+                return "-"
+            response_dict = json.loads(response.get("ResourceDescription", {}).get("Properties", {}))
+            furl = response_dict.get("FunctionUrl", "-")
+            return furl
+        except ClientError as e:
+            if get_client_error_code(e) == "ResourceNotFoundException":
+                return "-"
+            LOG.error("ClientError Exception : %s", str(e))
+            raise SamListUnknownClientError(msg=str(e)) from e
+
+    def get_stage_list(self, api_id: str, api_type: APIGatewayEnum) -> List[Any]:
+        """
+        Gets a list of stages for a given api of type AWS::ApiGateway::RestApi or AWS::ApiGatewayV2::Api
+
+        Parameters
+        ----------
+        api_id: str
+            The api id or rest api id of the api
+        api_type: APIGatewayEnum
+            The type of api, AWS::ApiGateway::RestApi or AWS::ApiGatewayV2::Api
+
+        Returns
+        -------
+        response_list: List[Any]
+            A list of stages for the api
+        """
+        response_list: List[Any]
+        try:
+            response_list = []
+            response: dict
+            search_key: str
+            stage_name_key: str
+            if api_type == APIGatewayEnum.API_GATEWAY:
+                response = self.apigateway_client.get_stages(restApiId=api_id)
+                search_key = "item"
+                stage_name_key = "stageName"
+            else:
```

Comment:
can we make this a non-catch-all statement, and have actions per api_type, since its an enum it should have finite number of states.

Reply:
Changed it